### PR TITLE
📊 : WDI update

### PR DESCRIPTION
**Pull request by Marigold**
Update Worldbank WDI dataset. Previous version was loading variable metadata from grapher which wasn't optimal (introducing weird loop between ETL and MySQL). This version keeps metadata in YAML file and adds descriptions from the file published by Worldbank. You can find the [dataset on staging](http://staging-site-mojmir/admin/datasets/6038).

Notebook `update_metadata.ipynb` generates YAML file and adds missing sources (using ChatGPT recommendations). It can be reused on next WDI update.

I went through [WDI issues from the last update](https://github.com/owid/owid-issues/issues/1034) and checked them against the new version.

Chart revision tool has 500 charts... I'm going through them on staging, haven't found any issues yet. (I don't think this has to block review. I'll make sure to go through all of them before merging.)

## TODO
* [ ] chart revision tool
* [ ] archive dataset in grapher
* [ ] archive dataset in ETL

### CODE REVIEW

**Review by pabloarosado — APPROVED** [review]
Impressive! I haven't looked through the `wdi.meta.yml` and `wdi.sources.yml` files. I have made some minor suggestions, feel free to disregard them if it's too much hassle. And let me know if you need help with chart revisions.

File: etl/steps/data/garden/worldbank_wdi/2023-05-29/wdi.countries.json [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+  "Korea, Dem. People's Rep.": "North Korea",
+  "Korea, Rep.": "South Korea",
+  "Lao PDR": "Laos",
+  "Macao SAR, China": "Macao",
+  "St. Martin (French part)": "Saint Martin (French part)",
+  "Venezuela, RB": "Venezuela",
+  "West Bank and Gaza": "West Bank and Gaza",
+  "Yemen, Rep.": "Yemen",
+
+  "East Asia & Pacific": "East Asia and Pacific",
```
[the remark this answers is not in the document]
Comment (Marigold) [reply etl/steps/data/garden/worldbank_wdi/2023-05-29/wdi.countries.json:221]: You're right, we already do it for some of our datasets. I added that suffix everywhere.

File: etl/steps/data/garden/worldbank_wdi/2023-05-29/wdi.countries.json [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+  "Europe & Central Asia": "Europe and Central Asia",
+  "Latin America & Caribbean": "Latin America and Caribbean",
+  "Middle East & North Africa": "Middle East and North Africa",
+  "North America": "North America",
+  "South Asia": "South Asia",
+  "Sub-Saharan Africa": "Sub-Saharan Africa",
+
+  "Low income": "Low income",
```
[the remark this answers is not in the document]
Comment (Marigold) [reply etl/steps/data/garden/worldbank_wdi/2023-05-29/wdi.countries.json:229]: I didn't even notice. I used your definitions and excluded those two, it was definitely confusing for me. (By the way we're a bit inconsistent with our naming of income groups across datasets)

File: etl/steps/data/garden/worldbank_wdi/2023-05-29/wdi.py [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+
+    df = pd.DataFrame(tb_meadow)
+
+    df_cust = mk_custom_entities(df)
+    assert all([col in df.columns for col in df_cust.columns])
+    df = pd.concat([df, df_cust], axis=0)
+
+    tb_garden = Table(df, metadata=tb_meadow.metadata)
+
+    log.info("wdi.add_variable_metadata")
```
[the remark this answers is not in the document]
Comment (Marigold) [reply etl/steps/data/garden/worldbank_wdi/2023-05-29/wdi.py:48]: I don't think loading snapshots in garden steps is such a bad practice if there's only a little processing. We sometimes skip meadow steps (for backports, fast-track etc.)

File: etl/steps/data/garden/worldbank_wdi/2023-05-29/wdi.py [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+    """constructs observations for custom entities, to be appended to existing df
+
+    e.g. poverty headcount for "World (excluding China)"
+    """
+    # poverty headcount for world (excluding china)
+    pop_code = "sp_pop_totl"
+    pov_share_code = "si_pov_dday"
+    omm_pov_count_code = "omm_pov_count"
+    assert df[pov_share_code].min() >= 0.0 and df[pov_share_code].max() <= 100.0 and df[pov_share_code].max() > 1.0
+    s_temp = df[pop_code].astype(float).multiply(df[pov_share_code].divide(100)).round(0)
```
[the remark this answers is not in the document]
Comment (Marigold) [reply etl/steps/data/garden/worldbank_wdi/2023-05-29/wdi.py:316]: Yes, it is necessary here.

File: etl/steps/data/garden/worldbank_wdi/2023-05-29/wdi.py [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+        publisher_source=f"Our World in Data based on {source.published_by}",
+    )
+
+    # omm: urban population living in slums
+    urb_pop_code = "sp_urb_totl"  # Urban population
+    slums_pct_code = "en_pop_slum_ur_zs"  # Population living in slums (% of urban population)
+    omm_urb_pop_slum = "omm_urb_pop_slum"
+    omm_urb_pop_nonslum = "omm_urb_pop_nonslum"
+    assert df[slums_pct_code].min() >= 0.0 and df[slums_pct_code].max() <= 100.0 and df[slums_pct_code].max() > 1.0
+    tb_omm[omm_urb_pop_slum] = df[urb_pop_code].astype(float).multiply(df[slums_pct_code].divide(100)).round(0)
```
[the remark this answers is not in the document]
Comment (Marigold) [reply etl/steps/data/garden/worldbank_wdi/2023-05-29/wdi.py:104]: Same here. Percentages in WDI have like five decimals.

File: etl/steps/data/garden/worldbank_wdi/2023-05-29/wdi.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pabloarosado) [reply etl/steps/data/garden/worldbank_wdi/2023-05-29/wdi.py:319]: I don't think rounding is ever a good idea, unless you either have a value for the uncertainty or have a specific criterion on the number of significant figures (and we don't have either). I'd suggest to never round in garden. If anything, we can round in grapher, or simply set the number of decimals. No need to change it now (especially if this was also done in the previous version), but for the next time I think we should avoid rounding.

**Pull request merged**